File: api/src/main/java/io/github/chriszhong/model/organization/relation/manager/ModeratesManager.java
package io.github.chriszhong.model.organization.relation.manager;

import io.github.chriszhong.model.organization.entity.Attribute;
import io.github.chriszhong.model.organization.entity.Pmf;
import io.github.chriszhong.model.organization.id.UniqueId;
import io.github.chriszhong.model.organization.relation.Moderates;

import java.util.Set;

/**
 * The {@linkplain ModeratesManager} interface defines the APIs for managing {@linkplain Moderates}.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
public interface ModeratesManager {
	/**
	 * Creates a moderates relation between a {@linkplain Pmf} and an {@linkplain Attribute} in this {@linkplain ModeratesManager}.
	 *
	 * @param pmfId
	 *            the {@linkplain UniqueId} that represents the {@linkplain Pmf}.
	 * @param attributeId
	 *            the {@linkplain UniqueId} that represents the {@linkplain Attribute}.
	 */
	void addModerates(UniqueId<Pmf> pmfId, UniqueId<Attribute> attributeId);

	/**
	 * Returns an {@linkplain Attribute}, if it exists, that is moderated by a {@linkplain Pmf} in this {@linkplain ModeratesManager}.
	 *
	 * @param id
	 *            the {@linkplain UniqueId} that represents the {@linkplain Pmf}
	 * @return an {@linkplain Attribute} if it exists, <code>null</code> otherwise.
	 */
	Attribute getModerates(UniqueId<Pmf> id);

	/**
	 * Returns a set of {@linkplain Pmf}s that moderates an {@linkplain Attribute} in this {@linkplain ModeratesManager}.
	 *
	 * @param id
	 *            the {@linkplain UniqueId} that represents the {@linkplain Attribute}.
	 * @return a set of {@linkplain Pmf}s.
	 */
	Set<Pmf> getModeratedBy(UniqueId<Attribute> id);

	/**
	 * Removes a moderates relation between a {@linkplain Pmf} and ad {@linkplain Attribute} in this {@linkplain ModeratesManager}.
	 *
	 * @param pmfId
	 *            the {@linkplain UniqueId} that represents the {@linkplain Pmf}.
	 * @param attributeId
	 *            the {@linkplain UniqueId} that represents the {@linkplain Attribute}.
	 */
	void removeModerates(UniqueId<Pmf> pmfId, UniqueId<Attribute> attributeId);

	/**
	 * Removes all moderates relations from this {@linkplain ModeratesManager}.
	 */
	void removeAllModerates();
}

File: chazm-api/src/main/kotlin/runtimemodels/chazm/api/relation/ModeratesManager.kt
package runtimemodels.chazm.api.relation

import runtimemodels.chazm.api.entity.Attribute
import runtimemodels.chazm.api.entity.Pmf
import runtimemodels.chazm.api.id.AttributeId
import runtimemodels.chazm.api.id.PmfId

/**
 * The [ModeratesManager] interface defines the APIs for managing a set of [Moderates] relations.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
interface ModeratesManager {
    /**
     * Adds a [Moderates] relation between a [Pmf] and an [Attribute] to this [ModeratesManager].
     *
     * @param moderates       the [Moderates] to add.
     */
    fun add(moderates: Moderates)

    /**
     * Returns the [Moderates] relation between an [Pmf] and an [Attribute] from this [HasManager].
     *
     * @param pmfId           the [PmfId] that represents the [Pmf].
     * @param attributeId the [AttributeId] that represents the [Attribute].
     * @return the [Moderates] relation if it exists, `null` otherwise.
     */
    operator fun get(pmfId: PmfId, attributeId: AttributeId): Moderates?

    /**
     * Returns a [Map] of [Attribute]s that are contained by a [Pmf] from this [ModeratesManager].
     *
     * @param id the [PmfId] that represents the [Pmf].
     * @return a [Map] of [Attribute]s.
     */
    operator fun get(id: PmfId): Map<AttributeId, Moderates>

    /**
     * Returns a [Map] of [Pmf]s that contains a [Attribute] from this [ModeratesManager].
     *
     * @param id the [AttributeId] that represents the [Attribute].
     * @return a [Map] of [Pmf]s.
     */
    operator fun get(id: AttributeId): Map<PmfId, Moderates>

    /**
     * Removes a moderates relation between a [Pmf] and ad [Attribute] in this [ModeratesManager].
     *
     * @param pmfId       the [PmfId] that represents the [Pmf].
     * @param attributeId the [AttributeId] that represents the [Attribute].
     * @return the [Moderates] that was removed, `null` otherwise.
     */
    fun remove(pmfId: PmfId, attributeId: AttributeId): Moderates?

    /**
     * Removes all [Moderates] relations associated to a [Pmf] from this [ModeratesManager].
     *
     * @param id the [PmfId] that represents the [Pmf].
     */
    fun remove(id: PmfId)

    /**
     * Removes all [Moderates] relations associated to a [Attribute] from this [ModeratesManager].
     *
     * @param id the [AttributeId] that represents the [Attribute].
     */
    fun remove(id: AttributeId)

}
